lib/pure/asyncmacro.nim

```diff
@@ -121,14 +121,36 @@ proc verifyReturnType(typeName: string, node: NimNode = nil) =
     error("Expected return type of 'Future' got '$1'" %
           typeName, node)
 
+macro isInAsync(n: FutureBase): bool =
+  ## Used to provide more useful error messages with 'await'.
+  var owner = n.owner
+  # We're recursing until we find something which could cause `yield FutureBase` to fail
+  while owner.symKind notin {nskProc, nskFunc, nskIterator, nskMacro} and owner.owner != nil:
+    owner = owner.owner
+
+  result = newLit(false)
+
+  # If async, owner should be an iterator returning owned[FutureBase]
+  if owner.symKind == nskIterator:
+    let returnType = owner.getImpl().params[0]
+    if returnType.kind in {nnkCall, nnkCommand, nnkBracketExpr}:
+      let returnsFuture =
+        (returnType[0].eqIdent"owned" and returnType[1].eqIdent"FutureBase") or
+        returnType[0].eqIdent"FutureBase"
+
+      result = newLit(returnsFuture)
+
 template await*(f: typed): untyped {.used.} =
   static:
     error "await expects Future[T], got " & $typeof(f)
 
 template await*[T](f: Future[T]): auto {.used.} =
   var internalTmpFuture: FutureBase = f
-  yield internalTmpFuture
-  (cast[typeof(f)](internalTmpFuture)).read()
+  when isInAsync(internalTmpFuture):
+    yield internalTmpFuture
+    (cast[typeof(f)](internalTmpFuture)).read()
+  else:
+    static: error "Can only 'await' inside a 'async' marked proc."
```

Comment:
```suggestion
    static: error "Can only 'await' inside a proc marked as 'async'."
```

Reply:
Should we be even more specific on the error message? Maybe mention "proc marked with the 'async' pragma" or is that unnecessarily verbose.

Reply:
Perhaps the message should be along the lines of `"Can only 'await' inside a proc marked as 'async'. If wanting to call a 'async' proc in non-async place use 'waitFor'"`